homeassistant/components/airthings_ble/config_flow.py

```diff
@@ -0,0 +1,144 @@
+"""Config flow for Airthings BlE integration."""
+
+from __future__ import annotations
+
+import dataclasses
+import logging
+from typing import Any
+
+from airthings_ble import AirthingsBluetoothDeviceData, AirthingsDevice
+import voluptuous as vol
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth import (
+    BluetoothServiceInfo,
+    async_discovered_service_info,
+)
+from homeassistant.config_entries import ConfigFlow
+from homeassistant.const import CONF_ADDRESS
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class Discovery:
+    """A discovered bluetooth device."""
+
+    name: str
+    discovery_info: BluetoothServiceInfo
+    device: AirthingsDevice
+
+
+def get_name(device: AirthingsDevice) -> str:
+    """Generate name with identifier for device."""
+    return f"{device.name} {device.identifier}"
+
+
+class AirthingsConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Airthings BLE."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._discovered_device: Discovery | None = None
+        self._discovered_devices: dict[str, Discovery] = {}
+
+    async def _get_device_data(self, discovery_info: BluetoothServiceInfo):
+        ble_device = bluetooth.async_ble_device_from_address(
+            self.hass, discovery_info.address.upper()
+        )
+        if ble_device is None:
+            raise UpdateFailed(f"No device for {discovery_info.address} found")
+
+        airthings = AirthingsBluetoothDeviceData(_LOGGER)
+
+        try:
+            data = await airthings.update_device(ble_device)
+        except Exception as err:
+            raise UpdateFailed(f"Unable to fetch data: {err}") from err
```

Comment:
Please trapped the named exception and translate them to `errors` so the user gets feedback about what went wrong instead.

`UpdateFailed` isn't intended to be raise in a config flow

Reply:
Do you mean that I need to specify exceptions to catch on ln 62 as well? Or just raise the exceptions using `AbortFlow` like in switchbot config flow?

Reply:
Normally when a config flow fails you return an error that is shown to the user - look for things like this:

```
return self.async_abort(reason="not_supported")
```

The "reason" is a key into strings.json, which is translatable into the users native language. That means you can't use string substitution like you do here. Instead, you have to catch specific errors, and map those to specific translatable keys. So try and catch common exceptions for things like "connection failed" and then map those to strings.json keys (adding new ones if needed). You can have a catch all for unknown failures too.

It looks like `AbortFlow` does the same thing, though i've always factored my code to use the `return self.async_abort` form. You still need to give it a key that exists in your strings.json, though.

Reply:
Alright  :+1: added `cannot_connect` and catching `BleakError` instead of just Exceptions.